### ISSUE TITLE
RecursionError during linting 2

### ISSUE THREAD

**Issue by Raymi306** [pylint-dev/pylint#9393]
### Bug description

I first experienced the bug locally.

I then reproduced it in CI [here](https://github.com/Raymi306/python-tutorial/actions/runs/7683376000/job/20938697932?pr=25)

Commit link: https://github.com/Raymi306/python-tutorial/commit/af22bb6dc370267ad06d6a491d56127e2c1c67ea

### Configuration

```ini
[MASTER]

# A comma-separated list of package or module names from where C extensions may
# be loaded. Extensions are loading into the active Python interpreter and may
# run arbitrary code.
extension-pkg-allow-list=

# A comma-separated list of package or module names from where C extensions may
# be loaded. Extensions are loading into the active Python interpreter and may
# run arbitrary code. (This is an alternative name to extension-pkg-allow-list
# for backward compatibility.)
extension-pkg-whitelist=

# Return non-zero exit code if any of these messages/categories are detected,
# even if score is above --fail-under value. Syntax same as enable. Messages
# specified are enabled, while categories only check already-enabled messages.
fail-on=

# Specify a score threshold to be exceeded before program exits with error.
fail-under=10.0

# Files or directories to be skipped. They should be base names, not paths.
ignore=CVS

# Add files or directories matching the regex patterns to the ignore-list. The
# regex matches against paths.
ignore-paths=

# Files or directories matching the regex patterns are skipped. The regex
# matches against base names, not paths.
ignore-patterns=

# Python code to execute, usually for sys.path manipulation such as
# pygtk.require().
#init-hook=

# Use multiple processes to speed up Pylint. Specifying 0 will auto-detect the
# number of processors available to use.
jobs=0

# Control the amount of potential inferred values when inferring a single
# object. This can help the performance when dealing with large functions or
# complex, nested conditions.
limit-inference-results=100

# List of plugins (as comma separated values of python module names) to load,
# usually to register additional checkers.
load-plugins=

# Pickle collected data for later comparisons.
persistent=yes

# Min Python version to use for version dependend checks. Will default to the
# version used to run pylint.
py-version=3.9

# When enabled, pylint would attempt to guess common misconfiguration and emit
# user-friendly hints instead of false-positive error messages.
suggestion-mode=yes

# Allow loading of arbitrary C extensions. Extensions are imported into the
# active Python interpreter and may run arbitrary code.
unsafe-load-any-extension=no


[MESSAGES CONTROL]

# Only show warnings with the listed confidence levels. Leave empty to show
# all. Valid levels: HIGH, INFERENCE, INFERENCE_FAILURE, UNDEFINED.
confidence=

# Disable the message, report, category or checker with the given id(s). You
# can either give multiple identifiers separated by comma (,) or put this
# option multiple times (only on the command line, not in the configuration
# file where it should appear only once). You can also use "--disable=all" to
# disable everything first and then reenable specific checks. For example, if
# you want to run only the similarities checker, you can use "--disable=all
# --enable=similarities". If you want to run only the classes checker, but have
# no Warning level messages displayed, use "--disable=all --enable=classes
# --disable=W".
disable=raw-checker-failed,
        bad-inline-option,
        locally-disabled,
        file-ignored,
        suppressed-message,
        useless-suppression,
        deprecated-pragma,
        use-symbolic-message-instead

# Enable the message, report, category or checker with the given id(s). You can
# either give multiple identifier separated by comma (,) or put this option
# multiple time (only on the command line, not in the configuration file where
# it should appear only once). See also the "--disable" option for examples.
enable=c-extension-no-member


[REPORTS]

# Python expression which should return a score less than or equal to 10. You
# have access to the variables 'error', 'warning', 'refactor', and 'convention'
# which contain the number of messages in each category, as well as 'statement'
# which is the total number of statements analyzed. This score is used by the
# global evaluation report (RP0004).
evaluation=10.0 - ((float(5 * error + warning + refactor + convention) / statement) * 10)

# Template used to display messages. This is a python new-style format string
# used to format the message information. See doc for all details.
#msg-template=

# Set the output format. Available formats are text, parseable, colorized, json
# and msvs (visual studio). You can also give a reporter class, e.g.
# mypackage.mymodule.MyReporterClass.
output-format=text

# Tells whether to display a full report or only the messages.
reports=no

# Activate the evaluation score.
score=yes


[REFACTORING]

# Maximum number of nested blocks for function / method body
max-nested-blocks=5

# Complete name of functions that never returns. When checking for
# inconsistent-return-statements if a never returning function is called then
# it will be considered as an explicit return statement and no message will be
# printed.
never-returning-functions=sys.exit,argparse.parse_error


[STRING]

# This flag controls whether inconsistent-quotes generates a warning when the
# character used as a quote delimiter is used inconsistently within a module.
check-quote-consistency=no

# This flag controls whether the implicit-str-concat should generate a warning
# on implicit string concatenation in sequences defined over several lines.
check-str-concat-over-line-jumps=no


[MISCELLANEOUS]

# List of note tags to take in consideration, separated by a comma.
notes=FIXME,
      XXX,
      TODO

# Regular expression of note tags to take in consideration.
#notes-rgx=


[SPELLING]

# Limits count of emitted suggestions for spelling mistakes.
max-spelling-suggestions=4

# Spelling dictionary name. Available dictionaries: none. To make it work,
# install the 'python-enchant' package.
spelling-dict=

# List of comma separated words that should be considered directives if they
# appear and the beginning of a comment and should not be checked.
spelling-ignore-comment-directives=fmt: on,fmt: off,noqa:,noqa,nosec,isort:skip,mypy:

# List of comma separated words that should not be checked.
spelling-ignore-words=

# A path to a file that contains the private dictionary; one word per line.
spelling-private-dict-file=

# Tells whether to store unknown words to the private dictionary (see the
# --spelling-private-dict-file option) instead of raising a message.
spelling-store-unknown-words=no


[BASIC]

# Naming style matching correct argument names.
argument-naming-style=snake_case

# Regular expression matching correct argument names. Overrides argument-
# naming-style.
#argument-rgx=

# Naming style matching correct attribute names.
attr-naming-style=snake_case

# Regular expression matching correct attribute names. Overrides attr-naming-
# style.
#attr-rgx=

# Bad variable names which should always be refused, separated by a comma.
bad-names=foo,
          bar,
          baz,
          toto,
          tutu,
          tata

# Bad variable names regexes, separated by a comma. If names match any regex,
# they will always be refused
bad-names-rgxs=

# Naming style matching correct class attribute names.
class-attribute-naming-style=any

# Regular expression matching correct class attribute names. Overrides class-
# attribute-naming-style.
#class-attribute-rgx=

# Naming style matching correct class constant names.
class-const-naming-style=UPPER_CASE

# Regular expression matching correct class constant names. Overrides class-
# const-naming-style.
#class-const-rgx=

# Naming style matching correct class names.
class-naming-style=PascalCase

# Regular expression matching correct class names. Overrides class-naming-
# style.
#class-rgx=

# Naming style matching correct constant names.
const-naming-style=UPPER_CASE

# Regular expression matching correct constant names. Overrides const-naming-
# style.
#const-rgx=

# Minimum line length for functions/classes that require docstrings, shorter
# ones are exempt.
docstring-min-length=-1

# Naming style matching correct function names.
function-naming-style=snake_case

# Regular expression matching correct function names. Overrides function-
# naming-style.
#function-rgx=

# Good variable names which should always be accepted, separated by a comma.
good-names=f,
           i,
           j,
           k,
           ex,
           Run,
           _

# Good variable names regexes, separated by a comma. If names match any regex,
# they will always be accepted
good-names-rgxs=

# Include a hint for the correct naming format with invalid-name.
include-naming-hint=no

# Naming style matching correct inline iteration names.
inlinevar-naming-style=any

# Regular expression matching correct inline iteration names. Overrides
# inlinevar-naming-style.
#inlinevar-rgx=

# Naming style matching correct method names.
method-naming-style=snake_case

# Regular expression matching correct method names. Overrides method-naming-
# style.
#method-rgx=

# Naming style matching correct module names.
module-naming-style=snake_case

# Regular expression matching correct module names. Overrides module-naming-
# style.
#module-rgx=

# Colon-delimited sets of names that determine each other's naming style when
# the name regexes allow several styles.
name-group=

# Regular expression which should only match function or class names that do
# not require a docstring.
no-docstring-rgx=^_

# List of decorators that produce properties, such as abc.abstractproperty. Add
# to this list to register other decorators that produce valid properties.
# These decorators are taken in consideration only for invalid-name.
property-classes=abc.abstractproperty

# Naming style matching correct variable names.
variable-naming-style=snake_case

# Regular expression matching correct variable names. Overrides variable-
# naming-style.
#variable-rgx=


[LOGGING]

# The type of string formatting that logging methods do. `old` means using %
# formatting, `new` is for `{}` formatting.
logging-format-style=old

# Logging modules to check that the string format arguments are in logging
# function parameter format.
logging-modules=logging


[VARIABLES]

# List of additional names supposed to be defined in builtins. Remember that
# you should avoid defining new builtins when possible.
additional-builtins=

# Tells whether unused global variables should be treated as a violation.
allow-global-unused-variables=yes

# List of names allowed to shadow builtins
allowed-redefined-builtins=

# List of strings which can identify a callback function by name. A callback
# name must start or end with one of those strings.
callbacks=cb_,
          _cb

# A regular expression matching the name of dummy variables (i.e. expected to
# not be used).
dummy-variables-rgx=_+$|(_[a-zA-Z0-9_]*[a-zA-Z0-9]+?$)|dummy|^ignored_|^unused_

# Argument names that match this expression will be ignored. Default to name
# with leading underscore.
ignored-argument-names=_.*|^ignored_|^unused_

# Tells whether we should check for unused import in __init__ files.
init-import=no

# List of qualified module names which can have objects that can redefine
# builtins.
redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io


[TYPECHECK]

# List of decorators that produce context managers, such as
# contextlib.contextmanager. Add to this list to register other decorators that
# produce valid context managers.
contextmanager-decorators=contextlib.contextmanager

# List of members which are set dynamically and missed by pylint inference
# system, and so shouldn't trigger E1101 when accessed. Python regular
# expressions are accepted.
generated-members=

# Tells whether missing members accessed in mixin class should be ignored. A
# mixin class is detected if its name ends with "mixin" (case insensitive).
ignore-mixin-members=yes

# Tells whether to warn about missing members when the owner of the attribute
# is inferred to be None.
ignore-none=yes

# This flag controls whether pylint should warn about no-member and similar
# checks whenever an opaque object is returned when inferring. The inference
# can return multiple potential results while evaluating a Python object, but
# some branches might not be evaluated, which results in partial inference. In
# that case, it might be useful to still emit no-member and other checks for
# the rest of the inferred objects.
ignore-on-opaque-inference=yes

# List of class names for which member attributes should not be checked (useful
# for classes with dynamically set attributes). This supports the use of
# qualified names.
ignored-classes=optparse.Values,thread._local,_thread._local

# List of module names for which member attributes should not be checked
# (useful for modules/projects where namespaces are manipulated during runtime
# and thus existing member attributes cannot be deduced by static analysis). It
# supports qualified module names, as well as Unix pattern matching.
ignored-modules=

# Show a hint with possible names when a member name was not found. The aspect
# of finding the hint is based on edit distance.
missing-member-hint=yes

# The minimum edit distance a name should have in order to be considered a
# similar match for a missing member name.
missing-member-hint-distance=1

# The total number of similar names that should be taken in consideration when
# showing a hint for a missing member.
missing-member-max-choices=1

# List of decorators that change the signature of a decorated function.
signature-mutators=


[FORMAT]

# Expected format of line ending, e.g. empty (any line ending), LF or CRLF.
expected-line-ending-format=

# Regexp for a line that is allowed to be longer than the limit.
ignore-long-lines=^\s*(# )?<?https?://\S+>?$

# Number of spaces of indent required inside a hanging or continued line.
indent-after-paren=4

# String used as indentation unit. This is usually "    " (4 spaces) or "\t" (1
# tab).
indent-string='    '

# Maximum number of characters on a single line.
max-line-length=100

# Maximum number of lines in a module.
max-module-lines=1000

# Allow the body of a class to be on the same line as the declaration if body
# contains single statement.
single-line-class-stmt=no

# Allow the body of an if to be on the same line as the test if there is no
# else.
single-line-if-stmt=no


[SIMILARITIES]

# Comments are removed from the similarity computation
ignore-comments=yes

# Docstrings are removed from the similarity computation
ignore-docstrings=yes

# Imports are removed from the similarity computation
ignore-imports=no

# Signatures are removed from the similarity computation
ignore-signatures=no

# Minimum lines number of a similarity.
min-similarity-lines=4


[DESIGN]

# List of qualified class names to ignore when counting class parents (see
# R0901)
ignored-parents=

# Maximum number of arguments for function / method.
max-args=5

# Maximum number of attributes for a class (see R0902).
max-attributes=7

# Maximum number of boolean expressions in an if statement (see R0916).
max-bool-expr=5

# Maximum number of branch for function / method body.
max-branches=12

# Maximum number of locals for function / method body.
max-locals=15

# Maximum number of parents for a class (see R0901).
max-parents=7

# Maximum number of public methods for a class (see R0904).
max-public-methods=20

# Maximum number of return / yield for function / method body.
max-returns=6

# Maximum number of statements in function / method body.
max-statements=50

# Minimum number of public methods for a class (see R0903).
min-public-methods=2


[IMPORTS]

# List of modules that can be imported at any level, not just the top level
# one.
allow-any-import-level=

# Allow wildcard imports from modules that define __all__.
allow-wildcard-with-all=no

# Analyse import fallback blocks. This can be used to support both Python 2 and
# 3 compatible code, which means that the block might have code that exists
# only in one or another interpreter, leading to false positives when analysed.
analyse-fallback-blocks=no

# Deprecated modules which should not be used, separated by a comma.
deprecated-modules=

# Output a graph (.gv or any supported image format) of external dependencies
# to the given file (report RP0402 must not be disabled).
ext-import-graph=

# Output a graph (.gv or any supported image format) of all (i.e. internal and
# external) dependencies to the given file (report RP0402 must not be
# disabled).
import-graph=

# Output a graph (.gv or any supported image format) of internal dependencies
# to the given file (report RP0402 must not be disabled).
int-import-graph=

# Force import order to recognize a module as part of the standard
# compatibility libraries.
known-standard-library=

# Force import order to recognize a module as part of a third party library.
known-third-party=enchant

# Couples of modules and preferred modules, separated by a comma.
preferred-modules=


[CLASSES]

# Warn about protected attribute access inside special methods
check-protected-access-in-special-methods=no

# List of method names used to declare (i.e. assign) instance attributes.
defining-attr-methods=__init__,
                      __new__,
                      setUp,
                      __post_init__

# List of member names, which should be excluded from the protected access
# warning.
exclude-protected=_asdict,
                  _fields,
                  _replace,
                  _source,
                  _make

# List of valid names for the first argument in a class method.
valid-classmethod-first-arg=cls

# List of valid names for the first argument in a metaclass class method.
valid-metaclass-classmethod-first-arg=cls


[EXCEPTIONS]

# Exceptions that will emit a warning when being caught. Defaults to
# "BaseException, Exception".
overgeneral-exceptions=builtins.BaseException,
                       builtins.Exception
```

### Command used

```shell
pylint app
```

### Pylint output

```sh
pylint app
Exception on node <If l.158 at 0x7c03f9053130> in file '/home/andrew/projects/personal/python-tutorial/app/pages/first_steps.py'
Traceback (most recent call last):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/utils.py", line 1359, in safe_infer
    value = next(infer_gen)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 1916, in _infer
    lhs = list(left_node.infer(context=context))
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 179, in _infer_stmts
    for inf in stmt.infer(context=context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 179, in _infer_stmts
    for inf in stmt.infer(context=context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 1765, in _infer
    yield from callee.infer_call_result(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  [Previous line repeated 929 more times]
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 328, in infer_call_result
    for node in self._proxied.igetattr("__call__", context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2492, in igetattr
    attributes = self.getattr(name, context, class_context=class_context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2407, in getattr
    result = [self.special_attributes.lookup(name)]
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/interpreter/objectmodel.py", line 132, in lookup
    return getattr(self, IMPL_PREFIX + name)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/interpreter/objectmodel.py", line 605, in attr___call__
    return self._instance.instantiate_class()
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2363, in instantiate_class
    if any(cls.name in EXCEPTION_BASE_CLASSES for cls in self.mro()):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2922, in mro
    return self._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2891, in _compute_mro
    inferred_bases = list(self._inferred_bases(context=context))
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2871, in _inferred_bases
    baseobj = next(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2871, in <genexpr>
    baseobj = next(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 144, in infer
    context = context.extra_context.get(self, context)
RecursionError: maximum recursion depth exceeded while calling a Python object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/base/basic_checker.py", line 288, in visit_if
    self._check_using_constant_test(node, node.test)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/base/basic_checker.py", line 335, in _check_using_constant_test
    inferred = utils.safe_infer(test)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/utils.py", line 1363, in safe_infer
    raise AstroidError from e
astroid.exceptions.AstroidError
concurrent.futures.process._RemoteTraceback: 
"""
Traceback (most recent call last):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/utils.py", line 1359, in safe_infer
    value = next(infer_gen)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 1916, in _infer
    lhs = list(left_node.infer(context=context))
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 179, in _infer_stmts
    for inf in stmt.infer(context=context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 179, in _infer_stmts
    for inf in stmt.infer(context=context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 169, in infer
    for i, result in enumerate(self._infer(context=context, **kwargs)):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 103, in inner
    yield from generator
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/decorators.py", line 49, in wrapped
    for res in _func(node, context, **kwargs):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_classes.py", line 1765, in _infer
    yield from callee.infer_call_result(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 331, in infer_call_result
    for res in node.infer_call_result(caller, context):
  [Previous line repeated 929 more times]
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/bases.py", line 328, in infer_call_result
    for node in self._proxied.igetattr("__call__", context):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2492, in igetattr
    attributes = self.getattr(name, context, class_context=class_context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2407, in getattr
    result = [self.special_attributes.lookup(name)]
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/interpreter/objectmodel.py", line 132, in lookup
    return getattr(self, IMPL_PREFIX + name)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/interpreter/objectmodel.py", line 605, in attr___call__
    return self._instance.instantiate_class()
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2363, in instantiate_class
    if any(cls.name in EXCEPTION_BASE_CLASSES for cls in self.mro()):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2922, in mro
    return self._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2898, in _compute_mro
    mro = base._compute_mro(context=context)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2891, in _compute_mro
    inferred_bases = list(self._inferred_bases(context=context))
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2871, in _inferred_bases
    baseobj = next(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/scoped_nodes/scoped_nodes.py", line 2871, in <genexpr>
    baseobj = next(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/astroid/nodes/node_ng.py", line 144, in infer
    context = context.extra_context.get(self, context)
RecursionError: maximum recursion depth exceeded while calling a Python object

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 831, in _check_file
    check_astroid_module(ast_node)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1017, in check_astroid_module
    retval = self._check_astroid_module(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 1069, in _check_astroid_module
    walker.walk(node)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 94, in walk
    self.walk(child)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/utils/ast_walker.py", line 91, in walk
    callback(astroid)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/base/basic_checker.py", line 288, in visit_if
    self._check_using_constant_test(node, node.test)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/base/basic_checker.py", line 335, in _check_using_constant_test
    inferred = utils.safe_infer(test)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/checkers/utils.py", line 1363, in safe_infer
    raise AstroidError from e
astroid.exceptions.AstroidError

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/lib/python3.10/concurrent/futures/process.py", line 246, in _process_worker
    r = call_item.fn(*call_item.args, **call_item.kwargs)
  File "/usr/lib/python3.10/concurrent/futures/process.py", line 205, in _process_chunk
    return [fn(*args) for args in chunk]
  File "/usr/lib/python3.10/concurrent/futures/process.py", line 205, in <listcomp>
    return [fn(*args) for args in chunk]
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/parallel.py", line 79, in _worker_check_single_file
    _worker_linter.check_single_file_item(file_item)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 740, in check_single_file_item
    self._check_file(self.get_ast, check_astroid_module, file)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 833, in _check_file
    raise astroid.AstroidError from e
astroid.exceptions.AstroidError
"""

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/home/andrew/projects/personal/python-tutorial/venv/bin/pylint", line 8, in <module>
    sys.exit(run_pylint())
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/__init__.py", line 34, in run_pylint
    PylintRun(argv or sys.argv[1:])
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/run.py", line 211, in __init__
    linter.check(args)
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/pylinter.py", line 679, in check
    check_parallel(
  File "/home/andrew/projects/personal/python-tutorial/venv/lib/python3.10/site-packages/pylint/lint/parallel.py", line 153, in check_parallel
    for (
  File "/usr/lib/python3.10/concurrent/futures/process.py", line 575, in _chain_from_iterable_of_lists
    for element in iterable:
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 621, in result_iterator
    yield _result_or_cancel(fs.pop())
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 319, in _result_or_cancel
    return fut.result(timeout)
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 451, in result
    return self.__get_result()
  File "/usr/lib/python3.10/concurrent/futures/_base.py", line 403, in __get_result
    raise self._exception
astroid.exceptions.AstroidError
make: *** [Makefile:10: lint] Error 1
```

### Expected behavior

That pylint does not crash

### Pylint version

3.0.0 and 3.0.3 both failed for me, 2.17.7 did not

### OS / Environment

```
NAME="Pop!_OS"
VERSION="22.04 LTS"
ID_LIKE="ubuntu debian"
UBUNTU_CODENAME=jammy
```

Default terminal

python3.10

**Comment by jacobtylerwalls** [comment]
Thanks for the report. We're going to audit pylint's entrypoints to astroid's inference system (infer, safe_infer, igetattr, etc.) and start catching RecursionError. Will handle in #9139
